In [2]:
import pandas as pd

### Comparing the properties between ERDDAP servers (ARGO & EMSO ERDDAP servers)

Comparing metadata of each server's 'allDatasets' dataset

In [4]:
# Function to summarize DataFrame
def summarize_dataframe(df):
    summary = []
    for col in df.columns:
        col_summary = {
            'Column': col,
            'Data Type': df[col].dtype,
            'Number of Unique Values': df[col].nunique(),
            'Number of Missing Values': df[col].isna().sum(),
            'Sample Unique Values': df[col].unique()[:5]  # List first 5 unique values as sample
        }
        summary.append(col_summary)
    
    summary_df = pd.DataFrame(summary)
    return summary_df

##### following to be updated: 'overview_metadata' needs to retrieve from ERDDAP servers

In [8]:
# Load ARGO general metadata 
ARGO_general_metadata = pd.read_csv("files_euroargo_erddap/overview_metadata.csv")
# Load EMSO general metadata
EMSO_general_metadata = pd.read_csv("files_emso_erddap/overview_metadata.csv")

In [9]:
if (ARGO_general_metadata.columns == EMSO_general_metadata.columns).all():
    print("'allDatasets' dataset of both ERDDAP servers contains the same metadata properties")
else:
    print("'allDatasets' dataset of both ERDDAP servers contains different metadata properties")

'allDatasets' dataset of both ERDDAP servers contains the same metadata properties


In [10]:
# Get summary
ARGO_summary_df = summarize_dataframe(ARGO_general_metadata)
ARGO_summary_df

,Column,Data Type,Number of Unique Values,Number of Missing Values,Sample Unique Values
0,datasetID,object,54,0,"[allDatasets, OACP-Argo-NATL, OACP-Argo-Global..."
1,accessible,object,2,0,"[public, log in]"
2,institution,object,13,0,"[Ifremer, Ifremer/LOPS, LOPS/Ifremer, Argo, Je..."
3,dataStructure,object,2,0,"[table, grid]"
4,cdm_data_type,object,7,0,"[Other, Grid, TrajectoryProfile, Point, Profile]"
5,class,object,7,0,"[EDDTableFromAllDatasets, EDDTableFromMultidim..."
6,title,object,54,0,[* The List of All Active Datasets in this ERD...
7,minLongitude (degrees_east),float64,24,6,"[nan, -78.378, -242.156, -179.9998, -999.999]"
8,maxLongitude (degrees_east),float64,22,6,"[nan, -2.378, 129.844, 180.0, 181.706]"
9,longitudeSpacing (degrees_east),float64,9,22,"[nan, 1.0, 0.01, 0.1652301818181818, 0.0833338..."


In [11]:
# Get summary
EMSO_summary_df = summarize_dataframe(EMSO_general_metadata)
EMSO_summary_df

,Column,Data Type,Number of Unique Values,Number of Missing Values,Sample Unique Values
0,datasetID,object,322,0,"[allDatasets, EMSO_OBSEA_Buoy_Airmar_150WX_30m..."
1,accessible,object,1,0,[public]
2,institution,object,28,0,"[Axiom Docker Install, Polytechnic University ..."
3,dataStructure,object,2,0,"[table, grid]"
4,cdm_data_type,object,4,0,"[Other, Point, TimeSeries, Profile]"
5,class,object,3,0,"[EDDTableFromAllDatasets, EDDTableFromErddap, ..."
6,title,object,285,0,[* The List of All Active Datasets in this ERD...
7,minLongitude (degrees_east),float64,60,154,"[nan, 1.75257, 18.0824167, -8.927083, -32.275]"
8,maxLongitude (degrees_east),float64,60,154,"[nan, 1.75257, 18.0824167, -8.927083, -32.275]"
9,longitudeSpacing (degrees_east),float64,0,322,[nan]


Comparing metadata of each server's datasets

In [12]:
# Load ARGO properties from ARGO ERDDAP server
ARGO_props = pd.read_csv("files_euroargo_erddap/all_dataset_metadata.csv")
# Load ARGO properties from EMSO ERDDAP server
EMSO_props = pd.read_csv("files_emso_erddap/all_dataset_metadata.csv")

Comparing Dimension properties

In [5]:
ARGO_dimensions = pd.read_csv("files_euroargo_erddap/dataset_metadata_dimensions.csv")
# alternative, same results:
#ARGO_dimensions = ARGO_props[ ARGO_props['Row Type'] == "dimension" ]

ARGO_dimensions = ARGO_dimensions.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_dimensions

,Variable Name,Count,Values_list
0,PRES,1,[OS_DYFAMED_1994-2014_D_TSO2]
1,depth,28,"[HF_ac49_84ad_3eb6, SDC_BAL_CLIM_TS_V1_m, SDC_..."
2,frame,1,[traj_lonlat_init_t0]
3,latitude,32,"[HF_ac49_84ad_3eb6, OACP-Argo-Global, SDC_BAL_..."
4,longitude,32,"[HF_ac49_84ad_3eb6, OACP-Argo-Global, SDC_BAL_..."
5,ntraj,1,[traj_lonlat_init_t0]
6,time,31,"[HF_ac49_84ad_3eb6, OS_DYFAMED_1994-2014_D_TSO..."


In [6]:
EMSO_dimensions = pd.read_csv("files_emso_erddap/dataset_metadata_dimensions.csv")
# alternative, same results:
#EMSO_dimensions = EMSO_props[ EMSO_props['Row Type'] == "dimension" ]

EMSO_dimensions = EMSO_dimensions.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_dimensions

,Variable Name,Count,Values_list
0,depth,93,[Canary_Islands_OS_ESTOC_201510_D_CTD-FixO3-10...
1,time,140,[Canary_Islands_OS_ESTOC_201510_D_CTD-FixO3-10...


In [7]:
# Dimension properties common between both ERDDAP servers
common_dimensions = ARGO_dimensions[ARGO_dimensions['Variable Name'].isin(EMSO_dimensions['Variable Name'])]
common_dimensions = common_dimensions['Variable Name'].drop_duplicates()
common_dimensions

1    depth
6     time
Name: Variable Name, dtype: object

In [8]:
# Dimension properties unique to ARGO ERDDAP server
dimensions_unique_to_ARGO = ARGO_dimensions[~ARGO_dimensions['Variable Name'].isin(EMSO_dimensions['Variable Name'])]
dimensions_unique_to_ARGO = dimensions_unique_to_ARGO['Variable Name'].drop_duplicates()
dimensions_unique_to_ARGO

0         PRES
2        frame
3     latitude
4    longitude
5        ntraj
Name: Variable Name, dtype: object

In [9]:
# Dimension properties unique to EMSO ERDDAP server
dimensions_unique_to_EMSO = EMSO_dimensions[~EMSO_dimensions['Variable Name'].isin(ARGO_dimensions['Variable Name'])]
dimensions_unique_to_EMSO = dimensions_unique_to_EMSO['Variable Name'].drop_duplicates()
dimensions_unique_to_EMSO

Series([], Name: Variable Name, dtype: object)

In [10]:
# overview of compared Dimension properties
props_dim_summ = pd.DataFrame.from_dict({
    "ARGO_dim" : [len(ARGO_dimensions)],
    "EMSO_dim" : [len(EMSO_dimensions)],
    "common_dim" : [len(common_dimensions)],
    "ARGO_unique_dim" : [len(dimensions_unique_to_ARGO)],
    "EMSO_unique_dim" : [len(dimensions_unique_to_EMSO)]
})
props_dim_summ

,ARGO_dim,EMSO_dim,common_dim,ARGO_unique_dim,EMSO_unique_dim
0,7,2,2,5,0


Comparing Variable properties

In [11]:
ARGO_variables = pd.read_csv("files_euroargo_erddap/dataset_metadata_variables.csv")
# alternative, same results:
#ARGO_variables = ARGO_props[ ARGO_props['Row Type'] == "variable" ]

ARGO_variables = ARGO_variables.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_variables

,Variable Name,Count,Values_list
0,ACTI,1,[PCM-Argo]
1,Access_ordering_of_data,1,[SDC_BLS_DATA_TS_V1]
2,Access_restriction,1,[SDC_BLS_DATA_TS_V1]
3,Alternative_cruise_name,1,[SDC_BLS_DATA_TS_V1]
4,Alternative_station_name,1,[SDC_BLS_DATA_TS_V1]
...,...,...,...
570,vertical_sampling_scheme,1,[ArgoFloats]
571,vn,2,"[drifter_6hour_qc, drifter_hourly_qc]"
572,wcs,1,[allDatasets]
573,wmo_inst_type,2,"[ArgoFloats, ArgoFloats-synthetic-BGC]"


In [12]:
EMSO_variables = pd.read_csv("files_emso_erddap/dataset_metadata_variables.csv")
# alternative, same results:
#EMSO_variables = EMSO_props[ EMSO_props['Row Type'] == "variable" ]

EMSO_variables = EMSO_variables.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_variables

,Variable Name,Count,Values_list
0,AHGT,2,"[BB_595_CTD, FF_734_CTD]"
1,AIRT,13,"[EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min, EMS..."
2,AIRT_QC,13,"[EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min, EMS..."
3,ALKY,2,"[EMSO_OBSEA_Besos_Buoy_SA8065_30min, EMSO_OBSE..."
4,ALKY_QC,2,"[EMSO_OBSEA_Besos_Buoy_SA8065_30min, EMSO_OBSE..."
...,...,...,...
473,velocity_upward,4,"[smartbay_obs_15min_mean, smartbay_obs_adcp, s..."
474,velocity_upward_qc,4,"[smartbay_obs_15min_mean, smartbay_obs_adcp, s..."
475,wcs,1,[allDatasets]
476,wms,1,[allDatasets]


In [13]:
# Variable properties common between both ERDDAP servers
common_variables = ARGO_variables[ARGO_variables['Variable Name'].isin(EMSO_variables['Variable Name'])]
common_variables = common_variables['Variable Name'].drop_duplicates()
common_variables

17                 CDOM
18              CDOM_QC
23                 CNDC
24              CNDC_QC
25     CNDC_UNCERTAINTY
             ...       
532                time
533         timeSpacing
536               title
572                 wcs
574                 wms
Name: Variable Name, Length: 63, dtype: object

In [14]:
# Variable properties unique to ARGO ERDDAP server
variables_unique_to_ARGO = ARGO_variables[~ARGO_variables['Variable Name'].isin(EMSO_variables['Variable Name'])]
variables_unique_to_ARGO = variables_unique_to_ARGO['Variable Name'].drop_duplicates()
variables_unique_to_ARGO

0                          ACTI
1       Access_ordering_of_data
2            Access_restriction
3       Alternative_cruise_name
4      Alternative_station_name
                 ...           
568           up_radiance555_qc
569                          ve
570    vertical_sampling_scheme
571                          vn
573               wmo_inst_type
Name: Variable Name, Length: 512, dtype: object

In [15]:
# Variable properties unique to EMSO ERDDAP server
variables_unique_to_EMSO = EMSO_variables[~EMSO_variables['Variable Name'].isin(ARGO_variables['Variable Name'])]
variables_unique_to_EMSO = variables_unique_to_EMSO['Variable Name'].drop_duplicates()
variables_unique_to_EMSO

0                    AHGT
1                    AIRT
2                 AIRT_QC
3                    ALKY
4                 ALKY_QC
              ...        
471        velocity_north
472     velocity_north_qc
473       velocity_upward
474    velocity_upward_qc
477                xyz_QC
Name: Variable Name, Length: 415, dtype: object

In [16]:
# overview of compared Variable properties
props_var_summ = pd.DataFrame.from_dict({
    "ARGO_var" : [len(ARGO_variables)],
    "EMSO_var" : [len(EMSO_variables)],
    "common_var" : [len(common_variables)],
    "ARGO_unique_var" : [len(variables_unique_to_ARGO)],
    "EMSO_unique_var" : [len(variables_unique_to_EMSO)]
})
props_var_summ

,ARGO_var,EMSO_var,common_var,ARGO_unique_var,EMSO_unique_var
0,575,478,63,512,415


Comparing attribute properties

In [17]:
ARGO_attributes = pd.read_csv("files_euroargo_erddap/dataset_metadata_attributes.csv")
# alternative, same results:
#ARGO_attributes = ARGO_props[ ARGO_props['Row Type'] == "attribute" ]

#ARGO_attributes.head()
ARGO_attributes = ARGO_attributes.groupby('Attribute Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_attributes

,Attribute Name,Count,Values_list
0,Acknowledgements,6,"[SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_..."
1,Author_e_mail,23,"[SDC_BLS_CLIM_TS_V1_m, SDC_BLS_CLIM_TS_V1_s, S..."
2,CDI,4,"[SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_..."
3,CDO,4,"[SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_..."
4,C_format,44,"[ArgoFloats, ArgoFloats, ArgoFloats, ArgoFloat..."
...,...,...,...
326,version,3,"[ArgoFloats-reference, OACP-Argo-Global, OACP-..."
327,w_surf_option,1,[ariane_trajectories_qualitative]
328,west_longitude,1,[gosud]
329,wmo_platform_code,2,"[OS_DYFAMED_1994-2014_D_TSO2, drifter_6hour_qc]"


In [18]:
EMSO_attributes = pd.read_csv("files_emso_erddap/dataset_metadata_attributes.csv")
# alternative, same results:
#EMSO_attributes = EMSO_props[ EMSO_props['Row Type'] == "attribute" ]

#EMSO_attributes.head()
EMSO_attributes = EMSO_attributes.groupby('Attribute Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_attributes

,Attribute Name,Count,Values_list
0,CITATION,4,"[BB_505_ADCP, BB_595_CTD, FF_644_ADCP, FF_734_..."
1,Conventions,419,"[BB_505_ADCP, BB_567_SBE56, BB_584_SBE56, BB_5..."
2,DM_indicator,72,"[Emso_Ligure_Dyfamed_SedimentTrap, Emso_Ligure..."
3,Easternmost_Easting,168,"[BB_505_ADCP, BB_567_SBE56, BB_584_SBE56, BB_5..."
4,Netcdf_version,12,"[OS_ESTOC-1_200311_D_CTD_ANIMATE-10m, OS_ESTOC..."
...,...,...,...
274,valid_range,10,[Emso_Western_Ligurian_Bathybot_ADCP_NetCDF_20...
275,variable_name,1029,[Canary_Islands_OS_ESTOC_201510_D_CTD-FixO3-10...
276,wfd_waterbody_name,9,"[smartbay_obs_15min_mean, smartbay_obs_acousti..."
277,wfd_waterbody_type,9,"[smartbay_obs_15min_mean, smartbay_obs_acousti..."


In [19]:
# Attribute properties common between both ERDDAP servers
common_attributes = ARGO_attributes[ARGO_attributes['Attribute Name'].isin(EMSO_attributes['Attribute Name'])]
common_attributes = common_attributes['Attribute Name'].drop_duplicates()
common_attributes

6                Conventions
13              DM_indicator
16       Easternmost_Easting
21     Northernmost_Northing
23              QC_procedure
               ...          
321                    units
322          update_interval
324                valid_max
325                valid_min
329        wmo_platform_code
Name: Attribute Name, Length: 136, dtype: object

In [20]:
# Attribute properties unique to ARGO ERDDAP server
attributes_unique_to_ARGO = ARGO_attributes[~ARGO_attributes['Attribute Name'].isin(EMSO_attributes['Attribute Name'])]
attributes_unique_to_ARGO = attributes_unique_to_ARGO['Attribute Name'].drop_duplicates()
attributes_unique_to_ARGO

0         Acknowledgements
1            Author_e_mail
2                      CDI
3                      CDO
4                 C_format
              ...         
323    user_manual_version
326                version
327          w_surf_option
328         west_longitude
330                 zsigma
Name: Attribute Name, Length: 195, dtype: object

In [21]:
# Attribute properties unique to EMSO ERDDAP server
attributes_unique_to_EMSO = EMSO_attributes[~EMSO_attributes['Attribute Name'].isin(ARGO_attributes['Attribute Name'])]
attributes_unique_to_EMSO = attributes_unique_to_EMSO['Attribute Name'].drop_duplicates()
attributes_unique_to_EMSO

0                  CITATION
4            Netcdf_version
6          Processing_level
7              QC_indicator
19     ancilliary_variables
               ...         
269             uncertainty
274             valid_range
275           variable_name
276      wfd_waterbody_name
277      wfd_waterbody_type
Name: Attribute Name, Length: 143, dtype: object

In [22]:
# overview of compared Attribute properties
props_attr_summ = pd.DataFrame.from_dict({
    "ARGO_attr" : [len(ARGO_attributes)],
    "EMSO_attr" : [len(EMSO_attributes)],
    "common_attr" : [len(common_attributes)],
    "ARGO_unique_attr" : [len(attributes_unique_to_ARGO)],
    "EMSO_unique_attr" : [len(attributes_unique_to_EMSO)]
})
props_attr_summ

,ARGO_attr,EMSO_attr,common_attr,ARGO_unique_attr,EMSO_unique_attr
0,331,279,136,195,143


In [23]:
props_var_summ

,ARGO_var,EMSO_var,common_var,ARGO_unique_var,EMSO_unique_var
0,575,478,63,512,415


In [24]:
props_dim_summ

,ARGO_dim,EMSO_dim,common_dim,ARGO_unique_dim,EMSO_unique_dim
0,7,2,2,5,0


Findings on comparison of properties of data between the ARGO and EMSO ERDDAP servers

- the ERDDAP server interface return data in a semi-standard manner
- 'allDatasets' dataset returns information about the datasets that are available through both servers. The metadata to describe this information is the same in both servers.  
- both ERDDAP servers return netDCF datasets: data properties can be divided into three categories: dimensions, variables and attributes 
- the data properties of datasets have some degree of overlap
    - 63 of variable properties 
    - 136 of attribute properties 
- but there is still large number of properties unique to each server:
    - 5 of dimension properties
    - 927 of variable properties
    - 338 of attribute properties
- properties were analysed based on name only, no information available about the properties meaning/semantics 
